File: problems/0066-plus-one/solution.cpp
#include <vector>
using namespace std;

class Solution {
public:
    /*
    Problem: 0066 - Plus One

    Given a non-empty array of decimal digits representing a non-negative integer,
    increment the integer by one and return the resulting array of digits.

    Approach:
    - Start from the least significant digit (rightmost).
    - If the digit is less than 9, increment it and return.
    - If the digit is 9, set it to 0 and carry 1 to the next digit.
    - After the loop, if we still have a carry, insert 1 at the front.

    Time Complexity: O(n), where n is the number of digits.
    Space Complexity: O(1) extra space (in-place), O(n) for the output vector.
    */
    vector<int> plusOne(vector<int>& digits) {
        int n = digits.size();

        // Traverse digits from right to left.
        for (int i = n - 1; i >= 0; --i) {
            if (digits[i] < 9) {
                digits[i] += 1;  // No carry needed.
                return digits;
            }
            digits[i] = 0;  // Carry over to the next digit.
        }

        // All digits were 9. Insert 1 at the front.
        digits.insert(digits.begin(), 1);
        return digits;
    }
};

File: problems/0066-plus-one/solution.py
class Solution:
    def plusOne(self, digits: List[int]) -> List[int]:
        """
        Problem: 0066 - Plus One

        Given a non-empty array of decimal digits representing a non-negative integer,
        increment the integer by one and return the resulting array of digits.

        Approach:
        - Start from the least significant digit and move left.
        - Add 1 to the last digit. If it becomes 10, set to 0 and carry the 1.
        - Repeat until no carry remains or we process all digits.
        - If carry remains at the end, insert 1 at the beginning.

        Time complexity: O(n), where n is the number of digits.
        Space complexity: O(1) extra space (in-place), O(n) for the output list.
        """
        n = len(digits)

        for i in range(n - 1, -1, -1):
            if digits[i] < 9:
                digits[i] += 1  # Simple case: just increment and return.
                return digits
            digits[i] = 0  # Set current digit to 0 and continue carry.

        # If loop finishes, all digits were 9. Add extra 1 at the beginning.
        return [1] + digits

